crates/relayer/src/keyring/ed25519_key_pair.rs

```diff
@@ -0,0 +1,120 @@
+
+    let seed = Seed::new(&mnemonic, "");
+
+    let base_key = ExtendedSecretKey::from_seed(seed.as_bytes())
+        .map_err(|err| Error::bip32_key_generation_failed(Ed25519KeyPair::KEY_TYPE, err.into()))?;
+
+    let private_key = base_key
+        .derive(&standard_path_to_derivation_path(hd_path))
+        .map_err(|err| Error::bip32_key_generation_failed(Ed25519KeyPair::KEY_TYPE, err.into()))?;
+
+    Ok(private_key)
+}
+
+fn standard_path_to_derivation_path(path: &StandardHDPath) -> DerivationPath {
+    let child_indices = vec![
+        ChildIndex::hardened(path.purpose().as_value().as_number())
+            .expect("Purpose is not Hardened"),
+        ChildIndex::hardened(path.coin_type()).expect("Coin Type is not Hardened"),
+        ChildIndex::hardened(path.account()).expect("Account is not Hardened"),
+        ChildIndex::normal(path.change()).expect("Change is Hardened"),
+        ChildIndex::normal(path.index()).expect("Index is Hardened"),
+    ];
+
+    DerivationPath::new(child_indices)
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Ed25519KeyPair {
+    keypair: Keypair,
+}
+
+impl Clone for Ed25519KeyPair {
+    fn clone(&self) -> Self {
+        let Self {
+            keypair: Keypair { secret, public },
+        } = self;
+        Self {
+            keypair: Keypair {
+                secret: SecretKey::from_bytes(secret.as_bytes()).unwrap(),
+                public: *public,
+            },
+        }
+    }
+}
+
+impl Ed25519KeyPair {
+    fn from_mnemonic_internal(mnemonic: &str, hd_path: &StandardHDPath) -> Result<Self, Error> {
+        let private_key = private_key_from_mnemonic(mnemonic, hd_path)?.secret_key;
+        let public_key = (&private_key).into();
+        let keypair = Keypair {
+            secret: private_key,
+            public: public_key,
+        };
+
+        Ok(Self { keypair })
+    }
+}
+
+impl SigningKeyPair for Ed25519KeyPair {
+    const KEY_TYPE: KeyType = KeyType::Ed25519;
+
+    fn from_key_file(key_file: KeyFile, hd_path: &StandardHDPath) -> Result<Self, Error> {
+        let key_pair = Self::from_mnemonic_internal(&key_file.mnemonic, hd_path)?;
+
+        let public_key_from_file = PublicKey::from_bytes(
+            &bs58::decode(key_file.pubkey)
+                .into_vec()
+                .map_err(Error::bs58_decode)?,
+        )
+        .map_err(Error::invalid_public_key)?;
+
+        if key_pair.keypair.public != public_key_from_file {
+            return Err(Error::public_key_mismatch(
+                key_pair.keypair.public.as_bytes().to_vec(),
+                public_key_from_file.as_bytes().to_vec(),
+            ));
+        }
+        Ok(key_pair)
+    }
+
+    fn from_mnemonic(
+        mnemonic: &str,
+        hd_path: &StandardHDPath,
+        _address_type: &AddressType,
+        _account_prefix: &str,
+    ) -> Result<Self, Error> {
+        Self::from_mnemonic_internal(mnemonic, hd_path)
+    }
+
+    // Solana address: base58(pubkey)
```

Comment:
What if chains other than Solana use a different encoding? It's not an issue since we don't even support Solana itself at the moment, much less other chains which may use Ed25519 keys, but it might worth thinking about we would solve this eventually. Just some food for thought, no need to address it in this PR.

Reply:
One way to go about this is similar to what we are doing for Cosmos vs Ethermint addresses, where we have a `Secp256k1AddressType` enum that we use to derive the account. Might be worth adding something similar for Ed25519 keys for future-proofing.

Reply:
Added an `Ed25519AddressType`.